### ISSUE TITLE
Improve Matching Performance

### ISSUE THREAD

**Issue by mmmries** [mmmries/bayesic#1]
I really think this library should be able to do at least 1k matches per second given a dataset of 10k possible classifications. It seems odd that the current master branch can only do ~440/sec.

Below are the results of running `MIX_ENV=prod mix run benchmarks/training_and_matching.exs benchmarks/capshare_account_names.csv benchmarks/matched_account_names.csv`

> The referenced CSV files are sample data from a production system so I can't commit them here, but they represent ~10k training strings and 194 strings attempting to match.

```
Name                     ips        average  deviation         median         99th %
training               17.62       56.77 ms    ±14.64%       55.00 ms       83.66 ms
match 194 rows          2.30      435.30 ms     ±1.59%      435.15 ms      451.00 ms
```

**Comment by mmmries** [comment]
I tried to attack this problem by writing a [port in go](https://github.com/mmmries/bayesic/compare/master...goport). The hypothesis here is that elixir is spending a lot of time generating new data structures because it is guaranteeing immutability. Since `go` has mutable data structures it should be __way__ faster. But surprisingly the metrics don't come out much better.

> Because this is using a port, benchee doesn't work cleanly for benchmarking. So I replicated the same types of benchmarks and tested them out.

```
Name                     ips         median
training                3.13      320.00 ms
match 194 rows          3.68      271.60 ms
```

I [left some notes](https://github.com/mmmries/bayesic/blob/9d3e145991e5ba7a823d58e0e2d16e1b6f85b20e/notes.md) for myself about how to make that branch work because I suspect I'm doing something wrong.

**Comment by mmmries** [comment]
I also tried optimizing this in the raw elixir. A quick profiler run told me that a lot of my time was spent in `Enum.count`. I thought Elixir was keeping a count in its data structures, but I guess it is actually traversing the whole data structure to check the size.

So I [created a branch](https://github.com/mmmries/bayesic/compare/61a845c...a963c44) which has a `finalize` step after you build up all of your training data. It just caches the counts around some of the sets of data and it stopped tracking the `tokens_by_classification`, because I don't actually need that data in order to calculate the probability.

```
Name                     ips        average  deviation         median         99th %
training               19.13       52.27 ms    ±13.29%       51.88 ms       69.45 ms
match 194 rows          6.79      147.32 ms     ±4.12%      146.30 ms      174.94 ms
```

This almost triples my matching rate compared to the master branch to `1,300+` matches per second.

I decided to also do some pruning on the data to remove common tokens [in this further commit](https://github.com/mmmries/bayesic/compare/61a845c...b3f59a4). This is a common optimization in Naive Bayes since it is only throwing away data that has is seen commonly in the data and therefore won't have a big impact to your probabilities.

__Pruning Tokens that Show up in 50%+ of Classes__

```
Name                     ips        average  deviation         median         99th %
training               17.66       56.62 ms    ±16.69%       55.73 ms       82.51 ms
match 194 rows          6.42      155.75 ms    ±15.40%      147.74 ms      253.87 ms
```

__Pruning Tokens that Show up in 20%+ of Classes__

```
Name                     ips        average  deviation         median         99th %
training               19.30       51.83 ms    ±15.87%       52.02 ms       78.78 ms
match 194 rows        132.83        7.53 ms    ±11.96%        7.26 ms       10.23 ms
```

__Pruning Tokens that Show up in 5%+ of Classes__

```
Name                     ips        average  deviation         median         99th %
match 194 rows        135.99        7.35 ms    ±11.50%        7.05 ms        9.68 ms
training               20.55       48.67 ms    ±14.60%       47.36 ms       64.46 ms
```

I also went back and measured the correct, unmatched and incorrectly matched instances in my set of 194 matching strings for the `master` branch and each of the different pruning levels and they all had the same accuracy:

* correct: 185
* incorrect: 1
* unmatched: 8

So finalizing + pruning gets me up to `26,382` matches per second (single threaded). This is the type of performance I was hoping for.
